src/gun.erl

```diff
@@ -1367,6 +1387,15 @@ connected_ws_only(Type, Event, State) ->
 %%
 %% @todo It might be better, internally, to pass around a URIMap
 %% containing the target URI, instead of separate Host/Port/PathWithQs.
+connected(cast, {ping, ReplyTo, StreamRef},
+		State=#state{protocol=Protocol, protocol_state=ProtoState}) ->
+	case erlang:function_exported(Protocol, ping, 3) of
+		true ->
+			Commands = Protocol:ping(ProtoState, dereference_stream_ref(StreamRef, State), ReplyTo),
+			commands(Commands, State);
+		false ->
+			ReplyTo ! {gun_error, self(), StreamRef, not_supported_for_protocol}
+	end;
```

Comment:
Better not have optional callbacks for protocols. Rather make them return an error, and check the return value.

Reply:
We have a `has_keepalive() -> boolean()` just to check if a protocol supports keepalive. Shall we follow this style and add `has_ping() -> boolean()`?

Reply:
I don't think they are different, the only difference will be whether they have been implemented, and that can be signified via the return value.

We need `has_keepalive` because we must start a timer. We don't need that for user pings, we can just call the function directly and propagate the error if any.

I think we should make a difference between protocols that will implement it in the future (http3 and ws are likely candidates, would be `not_implemented`) and protocols that won't (http, raw, socks... would be `unsupported_by_protocol` or something less verbose).

Reply:
They can just return `{error, not_implemented}` and `{error, unsupported_by_protocol}` and we don't even need to check the return value here. Then they'll be handled like any connection error and it leads to a disconnect. Does this seem OK?

Alternatively, I can wrap it in another tuple like `{ok, Commands} | {error, UnsupportedReason}`. I don't see the need for this now but I can do it if you prefer it.

Reply:
The error reasons seem OK to me.